src/Trains.NET.Rendering/LayerRenderer/TerrainRenderer.cs

```diff
@@ -0,0 +1,157 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Data;
+using System.Linq;
+using Trains.NET.Engine;
+using Trains.NET.Engine.Tracks;
+
+namespace Trains.NET.Rendering
+{
+    [Order(0)]
+    internal class TerrainRenderer : ILayerRenderer
+    {
+        private const int ContourHeight = 40;
+        private readonly ITerrainMap _terrainLayout;
+        private readonly IPixelMapper _pixelMapper;
+
+        public TerrainRenderer( ITerrainMap terrainLayout, IPixelMapper pixelMapper)
+        {
+            _terrainLayout = terrainLayout;
+            _pixelMapper = pixelMapper;
+        }
+
+        public bool Enabled { get; set; } = false;
+        public string Name => "Terrain";
+
+        public void Render(ICanvas canvas, int width, int height)
+        {
+            var contourLevels = GenerateListOfContourPointsForEachContourLevel();
+
+            foreach (var contourLevel in contourLevels.Keys)
+            {
+                var contourPoints = contourLevels[contourLevel];
+                DrawContourLine(canvas, contourPoints);
+            }
+        }
+
+        private void DrawContourLine(ICanvas canvas, List<ContourPoint> contourPoints)
+        {
+            foreach (var contourPoint in contourPoints)
+            {
+                var orderedList = OrderListByDistanceFromPoint(contourPoint, contourPoints);
+                var firstNearest = orderedList.Skip(1).First(); // The first element will be the original element since that will have distance zero
+                var secondNearest = orderedList.Skip(2).First();
```

Comment:
I don't like `First()` and I don't like `Single()` or `SingleOrDefault()` FYI

Reply:
Ok changed to FirstOrDefault() and also did a check to make sure orderedList has at least 3 elements.